### PR TITLE
upgrade Application facade from v6 to v7 to introduce Device support

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

Why is this change needed?

upgraded `Application` `facade` from `v6` to `v7` to introduce Device support

## QA steps

How do we verify that the change works?

nil

## Documentation changes

Does it affect current user workflow? CLI? API?

no

### CODE REVIEW

File: apiserver/facades/client/application/application.go [hunk 1]
```diff
@@ -101,13 +107,23 @@ func NewFacadeV5(ctx facade.Context) (*APIv5, error) {
...
-	api, err := newFacadeBase(ctx)
+	api, err := NewFacadeV7(ctx)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	return &APIv6{api}, nil
 }
 
+// NewFacadeV7 provides the signature required for facade registration
+// for versions 7.
```
Comment (wallyworld) [inline apiserver/facades/client/application/application.go]: "versions" should be "version".
Typo in original comment. Can fix all of them.

File: apiserver/facades/client/application/application.go [hunk 2]
```diff
@@ -228,6 +244,34 @@ func (api *APIv5) Deploy(args params.ApplicationsDeployV5) (params.ErrorResults,
 	return api.APIBase.Deploy(newArgs)
 }
 
+// Deploy fetches the charms from the charm store and deploys them
+// using the specified placement directives.
+// V6 deploy did not support policy, so pass through an empty string.
```
Comment (wallyworld) [inline apiserver/facades/client/application/application.go]: This comment about policy needs updating to talk about devices and empty map

File: apiserver/facades/client/application/application.go [hunk 3]
```diff
@@ -415,9 +459,11 @@ func deployApplication(
 		Constraints:       args.Constraints,
 		Placement:         args.Placement,
 		Storage:           args.Storage,
-		AttachStorage:     attachStorage,
-		EndpointBindings:  args.EndpointBindings,
-		Resources:         args.Resources,
+		// TODO(ycliuhw): current PR only includes facades version upgrade, inject `Device` into deeper logic is next step in sperate PR
```
Comment (wallyworld) [inline apiserver/facades/client/application/application.go]: We can still assign device here and just ignore it lower down

File: apiserver/facades/client/application/deploy.go [hunk 4]
```diff
@@ -36,6 +36,7 @@ type DeployApplicationParams struct {
 	// instead of a machine spec.
 	Placement        []*instance.Placement
 	Storage          map[string]storage.Constraints
+	Device           map[string]charm.Device
```
Comment (wallyworld) [inline apiserver/facades/client/application/deploy.go]: Devices

File: apiserver/params/params.go [hunk 5]
```diff
@@ -263,6 +264,7 @@ type ApplicationDeploy struct {
 	Placement        []*instance.Placement          `json:"placement,omitempty"`
 	Policy           string                         `json:"policy,omitempty"`
 	Storage          map[string]storage.Constraints `json:"storage,omitempty"`
+	Device           map[string]charm.Device        `json:"Device,omitempty"`
```
Comment (wallyworld) [inline apiserver/params/params.go]: Devices.
Also, json tag is lower case

**Comment by ycliuhw** [comment]
[add Device to Application Facade for juju deploy cmd and upgrade facade version](https://trello.com/c/NyAMRxXT/285-add-device-to-application-facade-for-juju-deploy-cmd-and-upgrade-facade-version)

File: apiserver/facades/client/application/application.go [hunk 6]
```diff
@@ -89,7 +94,8 @@ func NewFacadeV4(ctx facade.Context) (*APIv4, error) {
 	return &APIv4{api}, nil
 }
 
-// NewFacade provides the signature required for facade registration.
+// NewFacadeV5 provides the signature required for facade registration.
```
Comment (wallyworld) [inline apiserver/facades/client/application/application.go]: remove extra "."

File: apiserver/facades/client/application/deploy.go [hunk 7]
```diff
@@ -36,6 +36,7 @@ type DeployApplicationParams struct {
 	// instead of a machine spec.
 	Placement        []*instance.Placement
 	Storage          map[string]storage.Constraints
+	Devices          map[string]charm.Device
```
Comment (wallyworld) [inline apiserver/facades/client/application/deploy.go]: We shouldn't introduce charm artefacts here. As with storage, we need to introduce a new struct. See juju/juju/storage/constraints.go

File: apiserver/facades/client/application/get_test.go [hunk 8]
```diff
@@ -101,6 +101,65 @@ func (s *getSuite) TestClientApplicationGetSmoketestV5(c *gc.C) {
 	})
 }
 
+func (s *getSuite) TestClientApplicationGetSmoketestV6(c *gc.C) {
```
Comment (wallyworld) [inline apiserver/facades/client/application/get_test.go]: We don't need V6 and V7 tests here. The change in functionality for Get() was in V5. V6 abd V7 will use V5 Get()

File: apiserver/params/params.go [hunk 9]
```diff
@@ -263,6 +264,7 @@ type ApplicationDeploy struct {
 	Placement        []*instance.Placement          `json:"placement,omitempty"`
 	Policy           string                         `json:"policy,omitempty"`
 	Storage          map[string]storage.Constraints `json:"storage,omitempty"`
+	Devices          map[string]charm.Device        `json:"devices,omitempty"`
```
Comment (wallyworld) [inline apiserver/params/params.go]: As stated elsewhere, don't introduce a charm dependency here

File: apiserver/params/charms.go [hunk 10]
```diff
@@ -79,6 +79,15 @@ type CharmStorage struct {
 	Properties  []string `json:"properties,omitempty"`
 }
 
+// CharmDevice mirrors charm.Device.
+type CharmDevice struct {
+	Name        string `bson:"name"`
+	Description string `bson:"description"`
+	Type        string `bson:"type"`
+	CountMin    int64  `bson:"countmin"`
```
Comment (wallyworld) [inline apiserver/params/charms.go]: count-min etc
to be consistent with storage struct

File: apiserver/params/charms.go [hunk 11]
```diff
@@ -107,6 +116,7 @@ type CharmMeta struct {
 	Tags           []string                     `json:"tags,omitempty"`
 	Series         []string                     `json:"series,omitempty"`
 	Storage        map[string]CharmStorage      `json:"storage,omitempty"`
+	Device         map[string]CharmDevice       `json:"devices,omitempty"`
```
Comment (wallyworld) [inline apiserver/params/charms.go]: Devices

File: core/devices/constrains_test.go [hunk 12]
```diff
@@ -0,0 +1,4 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package devices_test
```
Comment (wallyworld) [inline core/devices/constrains_test.go]: There's a typo in the name of this file

File: core/devices/constraints.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+type Constraints struct {
+
+	// Type is the device type.
+	// currently supported types are
+	// - gpu
+	// - nvidia.com/gpu
+	// - amd.com/gpu
+	Type DeviceType `bson:"type"`
+
+	// Count is the number of devices that the charm requires.
```
Comment (wallyworld) [inline core/devices/constraints.go]: Count is the number of devices that the user has asked for - count min and max are the number of devices the charm requires

File: core/devices/constraints_test.go [hunk 14]
```diff
@@ -0,0 +1,4 @@
+// Copyright 2018 Canonical Ltd.
```
Comment (wallyworld) [inline core/devices/constraints_test.go]: No need to add empty test file

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**